File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
    Map<Integer,  Integer> countMap = new HashMap();
        for(int number : nums)
            countMap.put(number, countMap.getOrDefault(number,0) + 1);
        
        List<Integer>[] countsArray = new ArrayList[nums.length + 1];
        for(int key : countMap.keySet())
        {
            int count = countMap.get(key);
            if(countsArray[count] == null) countsArray[count] = new ArrayList();
            countsArray[count].add(key);
        }
        
        int[] result = new int[k];
        int index = 0;
        for(int i = nums.length; i >= 0; i--)
        {
            List<Integer> returnedArray = countsArray[i];
            if(returnedArray == null)
                continue;
            
            for(int numb : returnedArray)
            {
                result[index++] = numb;
                if(k == index) return result;
            }
            
        }
        
        return result;
        
    }
    
}

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        countArr = [[] for _ in range(len(nums) + 1)]
        frequency = {}
        for value in nums:
            frequency[value] = frequency.get(value, 0) + 1

        for num, frequency in frequency.items():
            countArr[frequency].append(num)

        result = []

        for bucket in reversed(countArr):
            for num in bucket:
                result.append(num)
                k -=1
                if(k == 0):
                    return result

        return result
        